[winpixevent] Added new port for DX12 PIX event headers

walbourn opened the pull request:
The DirectX 12 version of PIX uses a specific set of headers/libs/DLLs to do event mark-up. This adds an adapter port for the NuGet package that ships these headers.

> These supersede the existing `pix.h` files in the Windows SDK which were only used by the legacy Visual Studio Graphics Diagnostics (VGSD) legacy version of PIX for Direct3D 11.

walbourn commented:
Are these infrastructure failures?

walbourn commented:
@BillyONeal  Is there a concern with accepting this one, or does it need to wait for a review cycle?

walbourn commented:
Quick revision because someone noted the package uses MIT for the license. In the past it used a custom license, so I had it set to `null` in vcpkg.json. Now it should be "MIT".

WangWeiLin-MV commented:
Could you please provide a simple test case? My local test failed with `x64-windows\include\pix3.h(56,16): error C3646: 'FileName': unknown override specifier`

```cmake
cmake_minimum_required(VERSION 3.25.1)
project(Proj LANGUAGES CXX)

add_executable(main c.cpp)
target_compile_features(main PRIVATE cxx_std_23)

find_package(winpixevent CONFIG REQUIRED)
target_link_libraries(main PRIVATE Microsoft::WinPixEventRuntime)
```

```cpp
#include "pix3.h"
#include <iostream>
int main(void) {
    std::cout << "winpixevent" << std::endl;
    return 0;
}
```

walbourn commented:
Try adding `#include <d3d12.h>` or `#include <Windows.h>`

WangWeiLin-MV reviewed (approved):
The port usage tests pass with the following triplets:

* x64-windows
* x64-windows-static

WangWeiLin-MV commented:
> Try adding `#include <d3d12.h>` or `#include <Windows.h>` before pix3.h.

Thanks for the comment, it works.

JavierMatosD commented:
Thanks for the contribution! Since this library appears to be a runtime library that uses prebuilt binaries, I want to make sure this approach aligns with vcpkg’s policies. I’ll need to consult with the team about handling prebuilt binaries for this type of library.

I’ll get back to you once we’ve discussed it. Thanks for your patience!

walbourn commented:
OK. This is how *directx12-agility*, *dstorage*, *xaudio2redist*, and a number of other ports work. This port is an important port of our gaming story for DX12 as we push for developers to look at VCPKG rather than NuGet for their software components.

DominikGrabiec commented:
I have a CL which compiles the WinPixEventRuntime from source if an alternative solution is desired. I can work to submit a pull request.

walbourn commented:
> I have a CL which compiles the WinPixEventRuntime from source if an alternative solution is desired. I can work to submit a pull request.

If there's a GitHub repo for it, I can look at that.

walbourn commented:
I'll look at reworking it to use https://github.com/microsoft/PixEvents instead.

DominikGrabiec commented:
Last 2 commits from https://github.com/DominikGrabiec/vcpkg/tree/winpixeventruntime
This has worked in my local vcpkg install for a while already.
If you want I can submit a pull request for this.

walbourn commented:
Thanks, @DominikGrabiec . I am going to see if the owners will accept a CMake file directly to the repo rather than having to use `vcpkg-msbuild`, but I'll see if they are open to it. If nothing else, I can crib from your work here. Thanks!

DominikGrabiec commented:
The main thing that felt a bit weird to me was the fact that you ended up with 3 copies of the source tree, one original, one for debug, and one for release. This was also my first foray into developing a vcpkg port so I don't know if everything is optimal.

walbourn commented:
@BillyONeal  I'd like to get this in as it is using the NuGet package. I'll work with the owner of WinPixEventRuntime to get CMake files into their GitHub project and move this to using from-source instead. I don't think there's much use in building using MSBuild since that doesn't provide any real toolset agility like using clang.

Neumann-A commented:
> I don't think there's much use in building using MSBuild since that doesn't provide any real toolset agility like using clang.

hmm? You can set VCPKG_PLATFORM_TOOLSET to ClangCL? But of course cmake is more flexible regarding toolchains.

Bvsemmer commented:
Hey all,

Thank you for the work on this! Is anything preventing a merge still? Having this on vcpkg:master would be really helpful for me.

BillyONeal commented on ports/winpixevent/portfile.cmake:
```diff
@@ -0,0 +1,41 @@
+set(VCPKG_POLICY_DLLS_IN_STATIC_LIBRARY enabled)
```
What are the effects of `/MD` vs. `/MT` on folks using this? I observe that other DirectX things use this but it seems like it should be `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` if this thing links with the CRT, or `set(VCPKG_LIBRARY_LINAGE dynamic)` if it does not?

walbourn replied:
I can try `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` instead.

walbourn committed:
[winpixevent] Added new port for DX12 PIX event headers

walbourn committed:
Update baseline

walbourn commented:
@BillyONeal I rebased against the latest main and fixed the issue here.

I'm now using `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` and excludes ``staticcrt`` support.

walbourn commented:
Validated using:

```
.\vcpkg install winpixevent:x64-windows
.\vcpkg install winpixevent:x64-windows-static-md

.\vcpkg install winpixevent:arm64-windows
.\vcpkg install winpixevent:arm64-windows-static-md

.\vcpkg install winpixevent:x64-mingw-static
```

BillyONeal commented:
Thanks!
